fast_ms/src/diff/compress_ms.cpp

```diff
@@ -0,0 +1,913 @@
+				exit(1);
+			}
+		}
+		nZeros=0; i++;
+		permutedNZeros=0; j++;
+	}
+	if (verbose) cerr << "OK" << endl;
+}
+
+
+/**
+ * @return 1 iff $ms$ encodes at least one MS value that is at most $max$.
+ */
+uint8_t hasValueAtMost(sdsl::bit_vector &ms, size_type max) {
+	const size_type MS_SIZE = ms.size();
+	size_type i;
+	size_type nZeros, msValue;
+	
+	msValue=1; nZeros=0;
+	for (i=0; i<MS_SIZE; i++) {
+		if (ms[i]==0) {
+			nZeros++;
+			continue;
+		}
+		if (nZeros>0) msValue+=nZeros-1;
+		else msValue-=1;
+		if (msValue<=max) return 1;
+		nZeros=0;
+	}
+	return 0;
+}
+
+
+/**
+ * Prints all pairs $(nZeros,nOnes)$.
+ */
+void runPairsStats(sdsl::bit_vector &permutedMS) {
+	const size_type MS_SIZE = permutedMS.size();
+	size_type i;
+	size_type nZeros, nOnes;
+	
+	nZeros=1; nOnes=0;
+	for (i=1; i<MS_SIZE; i++) {
+		if (permutedMS[i]==0 && permutedMS[i-1]==1) {
+			cout << nZeros << "," << nOnes << endl;
+			nZeros=1; nOnes=0;
+			continue;
+		}
+		if (permutedMS[i]==0) nZeros++;
+		else nOnes++;
+	}
+	cout << nZeros << "," << nOnes << endl;
+}
+
+
+// ------------------------------- END OF PERMUTATION ------------------------------------
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+size_type compute_length(const size_type n_ones, const size_type n_zeros){
+    if (n_ones < n_zeros)
+        return 2 * (n_zeros - n_ones);
+    else if (n_ones < 2 * n_zeros)
+        return 2 * (n_ones - n_zeros) + 1;
+    return n_ones;
+}
+
+size_type comp2(const string ms_path, const InputFlags& flags) {
+    sdsl::int_vector_buffer<1> ms(ms_path, std::ios::in, 512);
+    size_type from = abs_point();
+    string c_ms_path = ms_path + ms_compression::to_str(flags.compression);
+    sdsl::int_vector_buffer<1> c_ms(c_ms_path, std::ios::out, 512);
+
+    size_type n_ones = 0, n_zeros = 0, out_idx=0;
+    for (size_type i=0; i<ms.size(); i++) {
+        if (ms[i] == 1) {
+            n_ones++;
+            continue;
+        }
+
+        assert(ms[i] ==  0);
+        if (n_ones == 0) {
+            n_zeros++;
+            c_ms[out_idx++] = 0;
+            continue;
+        }
+
+        size_type new_one_enc = compute_length(n_ones, n_zeros);
+        for(size_type k=0; k<new_one_enc; k++)
+            c_ms[out_idx++] = 1;
+        n_ones = 0;
+        c_ms[out_idx++] = 0;
+        n_zeros = 1;
+    }
+    if (n_ones > 0) {
+        size_type new_one_enc = compute_length(n_ones, n_zeros);
+        for(size_type k=0; k<new_one_enc; k++)
+            c_ms[out_idx++] = 1;
+    }
+    return diff_from(from);
+}
+
+
+template<typename vec_type, typename enc_type>
+size_type comp1(const string ms_path, const InputFlags& flags, const string outputDirectory, const string tablesDirectory, const uint32_t threshold, const size_type nZeros, const size_type nOnes, uint8_t allowNegativeMS, uint8_t greedyStrategy, uint8_t verbose) {
+	uint32_t initialMSValue;
+	sdsl::bit_vector ms;
+	
+cerr << "WTF> 1 \n";	
```

Comment:
are these error messages still informative to the user?

Reply:
oh, sorry! no, they should be removed. WTF.